### PR TITLE
Preserve `next-env.d.ts` line ending

### LINKED ISSUE #27324: Next 11.0.1 forces wrong line-endings on Windows in next-env.d.ts
### What version of Next.js are you using?

@latest v11.0.1

### What version of Node.js are you using?

@latest-ish v15.14.0

### What browser are you using?

Chrome

### What operating system are you using?

Windows 10 x64

### How are you deploying your application?

Vercel

### Describe the Bug

Next 11 writes `next-env.d.ts` all the time, but it uses *nix style line endings:
image
image

### Expected Behavior

*Don't manually force line-endings to CR
*Respect `.gitattributes` file (explictly specificing CR LF)
*Respect operating-system preference (CR LF on Windows)

### To Reproduce

1. Throw Prettier errors using this configuration:

`prettier.config.js`
```
module.exports = {
  // These settings are handled in .editorconfig:
  tabWidth: 2, // indent_size = 2
  useTabs: false, // indent_style = space
  endOfLine: "lf", // end_of_line = lf
  semi: false, // default: true
  singleQuote: false,
  printWidth: 80,
  trailingComma: "es5",
  bracketSpacing: true,
}
```

`.eslintrc.js`
```
module.exports = {
  extends: [
    "eslint:recommended",
    "plugin:prettier/recommended",
  ],
  rules: {
    "prettier/prettier": "error",
  },
}
```

### Preferable solution: Ask some of your engineers to switch to Windows full-time

### LINKED ISSUE #27980: LF to CRLF in `next-env.d.ts`
### What version of Next.js are you using?

11.1.0

### What version of Node.js are you using?

16.4.0

### What browser are you using?

does not matter

### What operating system are you using?

Windows

### How are you deploying your application?

next build

### Describe the Bug

For whatever reason, when running `next build` or starting a dev server with `next dev` or whatever, it always changes the `LF` line endings in the `next-env.d.ts` file to `CRLF` while making no other meaningful changes in it. That has a few further negative effects:
- the file gets marked as changed while obviously contains no meaningful changes (which is already irritating enough)
- your ESLint rules break on this file (if your setup expects LF only) which could be double-irritating if you have a pre-commit hook checking your project against ESLint
- if get around your pre-commit hook and your `git` is set up to transform any possible `CRLF` to `LF` automatically, then eventually you'll end-up with the proper format being pushed into repo. However once you re-launch the app locally, the story starts over again

Am I missing some important configuration detail in next.js? Or am I the only one affected by this?

### Expected Behavior

The file is NOT touched by `next.js` unless it has any meaningful changes to be applied there. Also, it should respect the line-endings that already exist in that file.

### To Reproduce

Please inquire if this is necessary, but running any setup on Windows having the file with LF endings should already be enough.

### PR DESCRIPTION
**Pull request by oluan**
Prevent next from changing already existing line ending on `next-env.d.ts`

- Fixes #27980
- Fixes #27324

### CODE REVIEW

File: packages/next/lib/typescript/writeAppTypeDeclarations.ts [hunk 1]
```diff
@@ -9,19 +9,39 @@ export async function writeAppTypeDeclarations(
...
+
+  try {
+    await fs.access(appTypeDeclarations, FS.F_OK)
+    const buffer = await fs.readFile(appTypeDeclarations)
+    const fileContent = buffer.toString()
+
+    // If file already exists then preserve its line ending
+    const lf = fileContent.indexOf('\n', /* skip first so we can lf - 1 */ 1)
+
+    if (lf != -1) {
```
Comment (styfle) [inline packages/next/lib/typescript/writeAppTypeDeclarations.ts:23]:
```suggestion
    if (lf !== -1) {
```

**Review by styfle — COMMENTED** [review]
Thanks!

Let's also add a test for this new behavior

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| buildDuration | 16.2s | 16.6s | ⚠️ +436ms |
| buildDurationCached | 4s | 4.1s | ⚠️ +69ms |
| nodeModulesSize | 61.5 MB | 61.5 MB | ⚠️ +1.58 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.066 | 3.144 | ⚠️ +0.08 |
| / avg req/sec | 815.32 | 795.24 | ⚠️ -20.08 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.874 | 1.807 | -0.07 |
| /error-in-render avg req/sec | 1333.91 | 1383.6 | +49.69 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| 745.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 23.2 kB | 23.2 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 67.1 kB | 67.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| _app-HASH.js gzip | 979 B | 979 B | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 329 B | 329 B | ✓ |
| dynamic-HASH.js gzip | 2.65 kB | 2.65 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 918 B | 918 B | ✓ |
| image-HASH.js gzip | 4.13 kB | 4.13 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 1.66 kB | 1.66 kB | ✓ |
| routerDirect..HASH.js gzip | 318 B | 318 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 320 B | 320 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 12.9 kB | 12.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| _buildManifest.js gzip | 492 B | 492 B | ✓ |
| Overall change | 492 B | 492 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| index.html gzip | 539 B | 539 B | ✓ |
| link.html gzip | 551 B | 551 B | ✓ |
| withRouter.html gzip | 532 B | 532 B | ✓ |
| Overall change | 1.62 kB | 1.62 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 4 Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| buildDuration | 14.1s | 13.6s | -455ms |
| buildDurationCached | 5.5s | 5.5s | -45ms |
| nodeModulesSize | 61.5 MB | 61.5 MB | ⚠️ +1.58 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.12 | 3.079 | -0.04 |
| / avg req/sec | 801.16 | 811.94 | +10.78 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.885 | 1.824 | -0.06 |
| /error-in-render avg req/sec | 1326.25 | 1370.26 | +44.01 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| 16.HASH.js gzip | 186 B | 186 B | ✓ |
| 677f882d2ed8..HASH.js gzip | 14.1 kB | 14.1 kB | ✓ |
| framework.HASH.js gzip | 41.9 kB | 41.9 kB | ✓ |
| main-HASH.js gzip | 10.6 kB | 10.6 kB | ✓ |
| webpack-HASH.js gzip | 1.19 kB | 1.19 kB | ✓ |
| Overall change | 68 kB | 68 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| _app-HASH.js gzip | 964 B | 964 B | ✓ |
| _error-HASH.js gzip | 3.8 kB | 3.8 kB | ✓ |
| amp-HASH.js gzip | 552 B | 552 B | ✓ |
| css-HASH.js gzip | 333 B | 333 B | ✓ |
| dynamic-HASH.js gzip | 2.85 kB | 2.85 kB | ✓ |
| head-HASH.js gzip | 3.06 kB | 3.06 kB | ✓ |
| hooks-HASH.js gzip | 924 B | 924 B | ✓ |
| index-HASH.js gzip | 231 B | 231 B | ✓ |
| link-HASH.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..HASH.js gzip | 298 B | 298 B | ✓ |
| script-HASH.js gzip | 2.98 kB | 2.98 kB | ✓ |
| withRouter-HASH.js gzip | 295 B | 295 B | ✓ |
| 30809af5c834..565.css gzip | 125 B | 125 B | ✓ |
| Overall change | 18.1 kB | 18.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| _buildManifest.js gzip | 500 B | 500 B | ✓ |
| Overall change | 500 B | 500 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| index.html gzip | 585 B | 585 B | ✓ |
| link.html gzip | 597 B | 597 B | ✓ |
| withRouter.html gzip | 578 B | 578 B | ✓ |
| Overall change | 1.76 kB | 1.76 kB | ✓ |

</details>

</details>
Commit: b2900e2e7731321aa8b746e6719cddc0081de5f6

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 7ceb8362ce318f13a4db1737f21aeecb3f3497a5

**test/unit/write-app-declarations.test.ts** 
- find config > should preserve CRLF EOL
- find config > should preserve LF EOL
- find config > should use OS EOL by default

<details>
<summary>Expand output</summary>

  ● find config › should preserve CRLF EOL

    ENOENT: no such file or directory, open '/home/runner/work/next.js/next.js/test/unit/fixtures/app-declarations/next-env.d.ts'



  ● find config › should preserve LF EOL

    ENOENT: no such file or directory, open '/home/runner/work/next.js/next.js/test/unit/fixtures/app-declarations/next-env.d.ts'



  ● find config › should use OS EOL by default

    ENOENT: no such file or directory, open '/home/runner/work/next.js/next.js/test/unit/fixtures/app-declarations/next-env.d.ts'



</details>

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| buildDuration | 15.3s | 15.5s | ⚠️ +213ms |
| buildDurationCached | 3.9s | 3.9s | -23ms |
| nodeModulesSize | 182 MB | 182 MB | ⚠️ +707 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.347 | 3.337 | -0.01 |
| / avg req/sec | 746.9 | 749.22 | +2.32 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.907 | 1.938 | ⚠️ +0.03 |
| /error-in-render avg req/sec | 1310.89 | 1290.01 | ⚠️ -20.88 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| 779.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 26.7 kB | 26.7 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 70.6 kB | 70.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| _app-HASH.js gzip | 977 B | 977 B | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 311 B | 311 B | ✓ |
| css-HASH.js gzip | 328 B | 328 B | ✓ |
| dynamic-HASH.js gzip | 2.67 kB | 2.67 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 918 B | 918 B | ✓ |
| image-HASH.js gzip | 4.14 kB | 4.14 kB | ✓ |
| index-HASH.js gzip | 260 B | 260 B | ✓ |
| link-HASH.js gzip | 1.66 kB | 1.66 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 13 kB | 13 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| _buildManifest.js gzip | 493 B | 493 B | ✓ |
| Overall change | 493 B | 493 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| index.html gzip | 540 B | 540 B | ✓ |
| link.html gzip | 552 B | 552 B | ✓ |
| withRouter.html gzip | 534 B | 534 B | ✓ |
| Overall change | 1.63 kB | 1.63 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| buildDuration | 8.1s | 7.9s | -161ms |
| buildDurationCached | 3.8s | 3.8s | -61ms |
| nodeModulesSize | 182 MB | 182 MB | ⚠️ +707 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.375 | 3.445 | ⚠️ +0.07 |
| / avg req/sec | 740.81 | 725.69 | ⚠️ -15.12 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.892 | 1.945 | ⚠️ +0.05 |
| /error-in-render avg req/sec | 1321.53 | 1285.65 | ⚠️ -35.88 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| 675-HASH.js gzip | 13.8 kB | 13.8 kB | ✓ |
| 770.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 50.7 kB | 50.7 kB | ✓ |
| main-HASH.js gzip | 34.6 kB | 34.6 kB | ✓ |
| webpack-HASH.js gzip | 1.65 kB | 1.65 kB | ✓ |
| Overall change | 101 kB | 101 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| polyfills-a4..dd70.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.32 kB | 1.32 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 315 B | 315 B | ✓ |
| css-HASH.js gzip | 331 B | 331 B | ✓ |
| dynamic-HASH.js gzip | 2.8 kB | 2.8 kB | ✓ |
| head-HASH.js gzip | 356 B | 356 B | ✓ |
| hooks-HASH.js gzip | 637 B | 637 B | ✓ |
| image-HASH.js gzip | 573 B | 573 B | ✓ |
| index-HASH.js gzip | 262 B | 262 B | ✓ |
| link-HASH.js gzip | 2.2 kB | 2.2 kB | ✓ |
| routerDirect..HASH.js gzip | 326 B | 326 B | ✓ |
| script-HASH.js gzip | 393 B | 393 B | ✓ |
| withRouter-HASH.js gzip | 322 B | 322 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 10.1 kB | 10.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| _buildManifest.js gzip | 511 B | 511 B | ✓ |
| Overall change | 511 B | 511 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | oluan/next.js next-env-line-ending | Change |
| - | - | - | - |
| index.html gzip | 539 B | 539 B | ✓ |
| link.html gzip | 551 B | 551 B | ✓ |
| withRouter.html gzip | 531 B | 531 B | ✓ |
| Overall change | 1.62 kB | 1.62 kB | ✓ |

</details>

</details>
Commit: 24581a6f6737dade951c5d870e828f8a47425a49

**Review by ijjk — APPROVED** [review]
Thanks for the PR!

**Pull request merged**